Product Collection: Add "Random" sorting option

imanish003 opened the pull request:
<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes #49530

This PR introduces a new 'Random' sorting option to the Product Collection block. These changes allow users to display products in a random order, enhancing the flexibility of product presentations in WooCommerce stores.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

1. Create a new post and add the Product Collection block.
2. Click on "create your own" button.
3. In inspector controls, verify that the "Random" sorting option is available in "Order By" dropdown.
4. Select the "Random" option and verify that the products are sorted in random order on Editor side.
5. Publish the post and verify that the products are sorted in random order on the frontend.
	- Refresh the page multiple times to ensure that the products are sorted in random order on the frontend.
6. Also verify that a few other sorting options are working as expected in this PR (test at least 2-3 of these):
   - A → Z
   - Z → A
   - Newest to oldest
   - Oldest to newest
   - Price, high to low
   - Price, low to high
   - Sales, high to low
   - Sales, low to high
   - Top Rated
   - Manual (menu order)

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->
Product Collection: Add "Random" sorting option
</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

imanish003 committed:
Add random sorting option to Product Collection block

This commit introduces a new 'Random' sorting option to the Product Collection block. The changes include:
1. Adding a new option for random sorting in the frontend UI (order-by-control.tsx).
2. Updating the backend to support the new random sorting option (ProductCollection.php).

These changes allow users to display products in a random order, enhancing the flexibility of product presentations in WooCommerce stores.

Commit:
Add changefile(s) from automation for the following project(s): woocommerce-blocks, woocommerce

imanish003 committed:
Add PHP unit tests for random sorting option

imanish003 committed:
Merge branch 'add/49530-product-collection-random-sorting-option' of https://github.com/woocommerce/woocommerce into add/49530-product-collection-random-sorting-option

octaedro commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/order-by-control.tsx:
```diff
@@ -63,6 +63,10 @@ const orderOptions = [
 		label: __( 'Manual (menu order)', 'woocommerce' ),
 	},
+	{
+		value: 'random/asc',
```
Is there a reason to add the `asc` after the `random`?

octaedro replied:
I don’t think `asc` or `desc` is needed for random order. It’s a minor detail, so feel free to keep it or remove it -it's up to you! 🙂

imanish003 replied:
It was needed because [here](https://github.com/woocommerce/woocommerce/blob/6fb3bab6a27668fbd71c202788c5f2ff36199629/plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/order-by-control.tsx#L100-L112) we assume that format will be like `orderBy/order` i.e. `date/asc` or `date/desc`. It will require some refactoring to remove extra "asc" from random sorting option. I was trying to avoid refactoring so as not to break anything. But let me check; it's probably easy to refactor it.

octaedro reviewed (approved):
Nice job @imanish003! This is testing well on my end and the code LGTM!
I'm going to pre-approve this PR. I just left a small question; other than that, the PR looks ready to go!

imanish003 committed:
Remove "asc" suffix from random sorting option

The "asc/desc" order is not needed for random sorting. This commit:
- Changes `random/asc` to just `random` in sorting options
- Updates code to handle sorting options without order suffix
- Adds fallback to 'asc' order when no order is specified

imanish003 commented:
Hey @octaedro, I have removed the "asc" suffix from the random sorting option and refactored the code to handle sorting options without an order suffix. Could you please review and test these changes to ensure everything works as expected? Let me know if you notice any issues.

imanish003 committed:
Fix: "order" shouldn't be present in REST API params

imanish003 committed:
Merge branch 'trunk' into add/49530-product-collection-random-sorting-option

octaedro commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/order-by-control.tsx:
```diff
@@ -75,7 +79,7 @@ const OrderByControl = ( props: QueryControlProps ) => {
 	};
 
-	let orderValue = `${ orderBy }/${ order }`;
+	let orderValue = order ? `${ orderBy }/${ order }` : orderBy;
```
❤️

octaedro reviewed (approved):
Thank you @imanish003, for addressing the suggestion. This PR looks almost ready to go! I noticed that [one action is still failing](https://github.com/woocommerce/woocommerce/actions/runs/11552602607/job/32156073408?pr=52322), but aside from that, everything looks fantastic. I'm pre-approving this PR ✅

octaedro commented:
I think the failing action should disappear [after updating the branch](https://github.com/woocommerce/woocommerce/pull/52379).

imanish003 committed:
Merge branch 'trunk' into add/49530-product-collection-random-sorting-option

asafm7 commented:
Will Random sorting work with pagination? And page caching?
